crypto/core_fetch.c

```diff
@@ -31,6 +31,20 @@ static void ossl_method_construct_this(OSSL_PROVIDER *provider,
     struct construct_data_st *data = cbdata;
     void *method = NULL;
 
+    /*
+     * Check to see if a method has already been constructed from this
+     * implementation.  Note that we use the property definition as a
+     * property query, which stresses that they MUST have compatible
+     * syntax, or else...  (in other words, it must always be possible
+     * to copy a property definition and paste that wherever a property
+     * is being queried.
```

Comment:
Sorry, but i don't understand the second part of this comment at all.
That may be an indication that I don't understand this code very well.

Reply:
You'll have to look at the manual for properties: https://github.com/openssl/openssl/blob/master/doc/man7/property.pod
The SYNTAX section is of interest here.  You have see that there's a `Definition` and a `Query` syntax, and they share certain properties, such as having `=` for assignment as well as for equality test.  This means that the *property definition* `foo=yes, bar=no` and the *property query* `foo=yes, bar=no` look exactly the same.  Now, the `put()` method takes a *property definition*, while the `get()` method takes a *property query*, but since the syntax is compatible, we can use the *property definition* as a *property query* to find the method that matches the exact \< name, properties \> pair that the provider gave us here.

If the *property query* syntax would change to use `==` for equality checks, this trick will no longer work, and we will need some other mechanism to "get" a method with an exact name and property definition.

@paulidale might have a word or two to say here...

Reply:
I rewrote the comment, in the hope that it's easier to understand

Reply:
Ah, interesting, that is...

Reply:
I'm not sure what to say.  It works and was designed this way, I don't see why it should ever not work (but I've been wrong before).

It would be possible to add a call to convert a definition to a query, it would be difficult to guarantee that the same algorithm was returned when the query was used.